### PR TITLE
Skip args in overloaded method names if all same

### LINKED ISSUE #712: Only mangle overloaded method names with the overload's extra argument names
```webidl
void foo(bool arg, optional any extra = null);
```

should generate methods named

1. `foo(arg: bool)`
2. `foo_using_extra(arg: bool, extra: &JsValue)`

but right now it generates

1. `foo_using_arg(arg: bool)`
2. `foo_using_arg_and_extra(arg: bool, extra: &JsValue)`

This is too wordy.

For a real world example, see `CanvasRenderingContext2D.prototype.arc`:

arc

+cc @afdw

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit updates how we name overloaded methods. Previously all argument
names were concatenated, but after this commit it only concatenates argument
names where at least one possibility has a different type. Otherwise if all
possibilities have the same type name it in theory isn't adding too much more
information!

Additionally this commit also switches to using `_with_` consistently everywhere
instead of `_with_` for constructors and `_using_` for methods.

Closes #712

### CODE REVIEW

File: examples/canvas/src/lib.rs [hunk 1]
```diff
@@ -31,43 +31,43 @@ pub fn draw() {
...
         50.0,
         0.0,
         f64::consts::PI * 2.0,
-    );
+    ).unwrap();
 
     // Draw the mouth.
     context.move_to(110.0, 75.0);
-    context.arc_using_x_and_y_and_radius_and_start_angle_and_end_angle(
+    context.arc(
```
Comment (fitzgen) [inline examples/canvas/src/lib.rs]: Yesssssssssss

**Comment by fitzgen** [comment]
Needs a rebase.

**Comment by afdw** [comment]
There was a reason why `using` was used. Methods were named like `original_name_with_override_argument_1_and_override_argument_2_using_possibility_argument_1_and_possibility_argument_2`. So `with` and `using` were used for different purposes: `with` for distinguishing overrides, `using` for distinguishing possibilities, both in constructors names and in method names. After this change many methods that have `with` twice in their names are generated:
```
add_ice_candidate_with_candidate_with_opt_rtc_ice_candidate
add_ice_candidate_with_candidate_with_opt_rtc_ice_candidate_init
append_with_name_and_value_and_filename_with_filename
clip_with_path_and_winding_with_winding
clip_with_winding_with_winding
connect_with_destination_and_output_and_input_with_output
connect_with_destination_and_output_and_input_with_output_and_input
connect_with_destination_and_output_with_output
create_answer_with_options_with_options
create_offer_with_options_with_options
draw_image_with_image_and_dx_and_dy_and_dw_and_dh_with_html_canvas_element
draw_image_with_image_and_dx_and_dy_and_dw_and_dh_with_html_image_element
draw_image_with_image_and_dx_and_dy_and_dw_and_dh_with_html_video_element
draw_image_with_image_and_dx_and_dy_and_dw_and_dh_with_image_bitmap
draw_image_with_image_and_dx_and_dy_and_dw_and_dh_with_svg_image_element
draw_image_with_image_and_dx_and_dy_with_html_canvas_element
draw_image_with_image_and_dx_and_dy_with_html_image_element
draw_image_with_image_and_dx_and_dy_with_html_video_element
draw_image_with_image_and_dx_and_dy_with_image_bitmap
draw_image_with_image_and_dx_and_dy_with_svg_image_element
draw_image_with_image_and_sx_and_sy_and_sw_and_sh_and_dx_and_dy_and_dw_and_dh_with_html_canvas_element
draw_image_with_image_and_sx_and_sy_and_sw_and_sh_and_dx_and_dy_and_dw_and_dh_with_html_image_element
draw_image_with_image_and_sx_and_sy_and_sw_and_sh_and_dx_and_dy_and_dw_and_dh_with_html_video_element
draw_image_with_image_and_sx_and_sy_and_sw_and_sh_and_dx_and_dy_and_dw_and_dh_with_image_bitmap
draw_image_with_image_and_sx_and_sy_and_sw_and_sh_and_dx_and_dy_and_dw_and_dh_with_svg_image_element
fill_with_path_and_winding_with_winding
fill_with_winding_with_winding
get_stats_with_selector_with_selector
is_point_in_path_with_path_and_x_and_y_and_winding_with_winding
is_point_in_path_with_x_and_y_and_winding_with_winding
open_with_method_and_url_and_async_and_user_and_password_with_user
open_with_method_and_url_and_async_and_user_and_password_with_user_and_password
open_with_name_and_options_with_options
open_with_type_and_replace_with_type
open_with_type_and_replace_with_type_and_replace
scroll_by_with_options_with_options
scroll_to_with_options_with_options
scroll_with_options_with_options
send_with_data_and_byte_offset_and_byte_length_with_byte_offset
send_with_data_and_byte_offset_and_byte_length_with_byte_offset_and_byte_length
set_interval_with_handler_and_timeout_and_unused_with_timeout_and_unused
set_timeout_with_handler_and_timeout_and_unused_with_timeout_and_unused
set_with_name_and_value_and_filename_with_filename
with_data_and_sw_and_sh_with_sh
with_node_and_output_and_options_with_output
with_node_and_output_and_options_with_output_and_options
with_p1_and_p2_and_p3_and_p4_with_p1
with_p1_and_p2_and_p3_and_p4_with_p1_and_p2
with_p1_and_p2_and_p3_and_p4_with_p1_and_p2_and_p3
with_p1_and_p2_and_p3_and_p4_with_p1_and_p2_and_p3_and_p4
with_params_with_params
with_stream_and_options_with_options
with_target_and_keyframes_and_options_with_opt_css_pseudo_element
with_target_and_keyframes_and_options_with_opt_css_pseudo_element_and_f64
with_target_and_keyframes_and_options_with_opt_css_pseudo_element_and_keyframe_effect_options
with_target_and_keyframes_and_options_with_opt_element
with_target_and_keyframes_and_options_with_opt_element_and_f64
with_target_and_keyframes_and_options_with_opt_element_and_keyframe_effect_options
```

**Comment by alexcrichton** [comment]
@afdw ah yes indeed! I think though that this is a bug in our naming that there's two locations for dealing with expanded function names, I'm working on a refactoring to treat everything a bit more uniformly
